sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala

```diff
@@ -651,4 +652,54 @@ class CastSuite extends CastSuiteBase {
         checkEvaluation(cast(cast(interval, StringType), YearMonthIntervalType()), period)
       }
   }
+
+  test("SPARK-35768: Take into account year-month interval fields in cast") {
+    checkEvaluation(cast(Literal.create("1-1"),
+      YearMonthIntervalType(YEAR, YEAR)), 12)
+    checkEvaluation(cast(Literal.create("1-1"),
+      YearMonthIntervalType(YEAR, MONTH)), 13)
+    checkEvaluation(cast(Literal.create("1-1"),
+      YearMonthIntervalType(MONTH, MONTH)), 13)
+    checkEvaluation(cast(Literal.create("-1-1"),
+      YearMonthIntervalType(YEAR, YEAR)), -12)
+    checkEvaluation(cast(Literal.create("-1-1"),
+      YearMonthIntervalType(YEAR, MONTH)), -13)
+    checkEvaluation(cast(Literal.create("-1-1"),
+      YearMonthIntervalType(MONTH, MONTH)), -13)
+
+    checkEvaluation(cast(Literal.create("INTERVAL '1-1' YEAR TO MONTH"),
```

Comment:
Could you rewrite this like a loop over a sequence of
(interval string, target type) -> <expected months>

Reply:
Done